File: client-go/internal/globals/globals.go
package globals

const MAX_PACKET_SIZE int = 4096

// Status Codes
const OK_STATUS_CODE uint8 = 0

// Top-level command byte, sent as the very first byte on every connection to
// spined's single socket ("/tmp/spine/spined") - mirrors
// protocol/src/globals.zig's REGISTER_NODE_CODE. spined and squid share that
// one socket now (ADD_NAMESPACE_CODE/GET_INFO_CODE are squid's own control
// commands on it); this tells spined the rest of the connection is a spine
// node registering itself (RegisterNodePayload follows).
const REGISTER_NODE_CODE uint8 = 6

// Entity-registration operation codes: the first byte of each registration
// message, telling spined which of RegisterPublisherPayload/
// RegisterServicePayload/RegisterConsumerPayload follows. Mirrors
// protocol/src/globals.zig's REGISTER_PUBLISHER_CODE/REGISTER_SERVICE_CODE/
// REGISTER_CONSUMER_CODE - three, not four, since a subscriber and a
// service_caller are both just a Consumer (name + which kind of producer
// it's waiting for) in spined's domain model now.
const REGISTER_PUBLISHER_CODE uint8 = 0
const REGISTER_SERVICE_CODE uint8 = 1
const REGISTER_CONSUMER_CODE uint8 = 2

// Rolls back a registration on the same connection - mirrors
// protocol/src/globals.zig's UNREGISTER_PUBLISHER_CODE/UNREGISTER_SERVICE_CODE/
// UNREGISTER_CONSUMER_CODE. Sent when a later step (binding the real
// listener, connecting to a producer) fails after spined already approved
// the registration, so spined doesn't keep believing an entity exists with
// nothing behind it. Idempotent on spined's side - unregistering a name
// that was never registered is not an error.
const UNREGISTER_PUBLISHER_CODE uint8 = 3
const UNREGISTER_SERVICE_CODE uint8 = 4
const UNREGISTER_CONSUMER_CODE uint8 = 5

const ERROR_SERIALIZER_ERROR_CODE uint8 = 251
const ERROR_SERVICE_ERROR_CODE uint8 = 252
const ERROR_MISMATCH_PAYLOAD_CODE uint8 = 254

// spined registration/entity status codes - mirrors protocol/src/globals.zig
// exactly, since these travel over the wire between spined and every client
// library, this one and client-zig alike.
const TOO_MANY_UNKNOWN_ENTITIES uint8 = 248
const NODE_ALREADY_REGISTERED uint8 = 249
const TOO_MANY_NODES uint8 = 250
const TOO_MANY_ENTITIES uint8 = 251
const INVALID_ENTITY_TYPE uint8 = 252
const ENTITY_ALREADY_REGISTERED uint8 = 253
const INVALID_NAMESPACE uint8 = 255

const ERROR_SERVICE_HANDLER = "service handler has an error"
const ERROR_CORRUPT_PAYLOAD = "CORRUPT_PAYLOAD"
const ERROR_PAYLOAD_SIZE = "failed to encode key. key is too big. max key size is 4kb"

File: protocol/src/globals.zig
// Shared spine wire protocol: the single Unix socket both spine nodes and
// squid dial to reach spined, the fixed string size used for every name on
// the wire, and every command/status/type code and array bound both sides
// must agree on numerically. mad encodes everything as fixed-size fields
// with no framing or length prefix, so a mismatch here is a silent wire
// break, not a compile error — squid, spined, and every spine client
// library all import this one copy instead of keeping their own in sync by
// hand.
pub const SPINED_PATH = "/tmp/spine/spined";
pub const PUBLISHER_SOCKET_DIR = "/tmp/spine/publisher/";
pub const SERVICE_SOCKET_DIR = "/tmp/spine/service/";

pub const STRING_SIZE: u32 = 32;

// Top-level command byte, sent as the very first byte on every SPINED_PATH
// connection, telling spined what the rest of the connection is:
//   - REGISTER_NODE_CODE: a spine node registering itself (RegisterNodePayload
//     follows, then any number of entity registration/unregistration
//     messages on the same connection - see namespace.zig's op codes below).
//   - ADD_NAMESPACE_CODE / GET_INFO_CODE: a one-shot squid control command.

pub const ADD_NAMESPACE_CODE: u8 = 0;
pub const REMOVE_NAMESPACE_CODE: u8 = 2;
pub const GET_INFO_CODE: u8 = 4;
pub const REGISTER_NODE_CODE: u8 = 6;

// squid control-command responses.
pub const OK_STATUS: u8 = 0;
pub const TOO_MANY_NAMESPACES: u8 = 240;
pub const NAMESPACE_ALREADY_REGISTERED: u8 = 241;
pub const INVALID_COMMAND: u8 = 246;
pub const ERROR_MISMATCH_PAYLOAD_CODE: u8 = 254;

pub fn statusMessage(status: u8) []const u8 {
    return switch (status) {
        OK_STATUS => "ok",
        TOO_MANY_NAMESPACES => "spined already has the maximum number of namespaces",
        NAMESPACE_ALREADY_REGISTERED => "a namespace with that name already exists",
        INVALID_COMMAND => "spined did not recognize that command",
        else => "unknown error",
    };
}

// GetInfoResponse array bounds. Fixed at compile time since mad has no
// dynamic-length encoding — these size the live storage in spined too, so
// nothing here can silently truncate a namespace/node/lookup list.
pub const MAX_NAMESPACES: u32 = 32;
pub const MAX_NODES: u32 = 256;
pub const MAX_ENTITIES: u32 = 256;
pub const MAX_SUBSCRIBERS_PER_PUBLISHER: usize = 32;

// Node/entity-registration protocol: spine nodes speak this to spined over
// SPINED_PATH (RegisterNodePayload, then any number of entity registration
// messages). Shared between spined and every spine client library so they
// can't drift the way spine-zig's own STRING_SIZE=64 copy once did.
pub const NODE_ALREADY_REGISTERED: u8 = 249;
pub const TOO_MANY_NODES: u8 = 250;
pub const INVALID_NAMESPACE: u8 = 255;

pub const TOO_MANY_UNKNOWN_ENTITIES: u8 = 248;
pub const TOO_MANY_ENTITIES: u8 = 251;
pub const INVALID_ENTITY_TYPE: u8 = 252;
pub const ENTITY_ALREADY_REGISTERED: u8 = 253;
pub const ERROR_SERIALIZER_ERROR_CODE: u8 = 251;
pub const ERROR_SERVICE_ERROR_CODE: u8 = 252;

// Entity-registration operation codes: the first byte of each registration
// message, telling spined which payload shape (RegisterPublisherPayload/
// RegisterServicePayload/RegisterConsumerPayload) follows - mirrors how
// squid's ADD_NAMESPACE_CODE/GET_INFO_CODE/etc. already work, rather than
// spined reading one fixed shape and branching on a field inside it.
//
// Three codes, not four: entity.zig's domain model already unifies
// subscriber/service_caller as one Consumer (name + which kind of producer
// it's waiting for), so the wire protocol mirrors that directly instead of
// carrying a 4th, mostly-empty payload shape for it.
pub const REGISTER_PUBLISHER_CODE: u8 = 0;
pub const REGISTER_SERVICE_CODE: u8 = 1;
pub const REGISTER_CONSUMER_CODE: u8 = 2;

// UNREGISTER_CONSUMER_CODE covers both subscriber and service_caller.
// Idempotent - unregistering a name that was never registered (or already
// removed) is OK_STATUS, not an error, since the caller's intent ("this
// shouldn't exist") is already satisfied either way.
pub const UNREGISTER_PUBLISHER_CODE: u8 = 3;
pub const UNREGISTER_SERVICE_CODE: u8 = 4;
pub const UNREGISTER_CONSUMER_CODE: u8 = 5;

pub const MAX_PACKET_SIZE: usize = 4096;

pub const REGISTER_ENTITY: u8 = 80;
pub const BAD_NAME: u8 = 254;

pub const MAX_FAKE_ENTITIES: u32 = 256;
pub const FAKE_NODE = "SPINED_FAKE_NODE";
